pkg/debuginfo/store.go

```diff
@@ -152,28 +160,33 @@ func (s *Store) Exists(ctx context.Context, req *debuginfopb.ExistsRequest) (*de
 		return nil, err
 	}
 
-	if found && req.Hash != "" {
-		dbgFile, err := s.fetchObjectFile(ctx, buildID)
+	if found {
+		var exists bool
+		metadataFile, err := s.metadataManager.fetch(ctx, buildID)
 		if err != nil {
+			if errors.Is(err, ErrMetadataNotFound) {
+				return &debuginfopb.ExistsResponse{Exists: false}, nil
+			}
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 
-		h, err := hash.File(dbgFile)
-		if err != nil {
-			return nil, status.Error(codes.Internal, err.Error())
+		if metadataFile.Hash == req.Hash {
+			exists = true
+			if metadataFile.State == metadataStateUploading {
```

Comment:
The hash is set once the file is persisted in `uploaded()` I think, so we don't have to check the state, as we know it will be `uploaded`, no?

Side note: perhaps checking for the state first can be clearer, like:

```
switch state:
  case metadataStateUploaded:
     // we know we should have a hash by now
  case metadataStateUploading:
     // we should check if the file is stale which means that it never completed uploading and the client should re-upload
...
```

What do you think?

Reply:
Good point. I think I've overlooked this.